src/main/java/net/minecraftforge/common/BiomeDictionary.java

```diff
@@ -249,251 +219,238 @@ public static boolean areBiomesEquivalent(Biome biomeA, Biome biomeB)
         {
-            BiomeDictionary.registerBiomeType(biome, SWAMP);
+            BiomeDictionary.addTypes(biome, SWAMP);
         }
 
         if (biome.getBaseHeight() <= -0.5F)
         {
             if (biome.getHeightVariation() == 0.0F)
             {
-                BiomeDictionary.registerBiomeType(biome, RIVER);
+                BiomeDictionary.addTypes(biome, RIVER);
             }
             else
             {
-                BiomeDictionary.registerBiomeType(biome, OCEAN);
+                BiomeDictionary.addTypes(biome, OCEAN);
             }
         }
 
         if (biome.getHeightVariation() >= 0.4F && biome.getHeightVariation() < 1.5F)
         {
-            BiomeDictionary.registerBiomeType(biome, HILLS);
+            BiomeDictionary.addTypes(biome, HILLS);
         }
 
         if (biome.getHeightVariation() >= 1.5F)
         {
-            BiomeDictionary.registerBiomeType(biome, MOUNTAIN);
+            BiomeDictionary.addTypes(biome, MOUNTAIN);
         }
 
         if (biome.getEnableSnow())
         {
-            BiomeDictionary.registerBiomeType(biome, SNOWY);
+            BiomeDictionary.addTypes(biome, SNOWY);
         }
 
         if (biome.topBlock != Blocks.SAND && biome.getTemperature() >= 1.0f && biome.getRainfall() < 0.2f)
         {
-            BiomeDictionary.registerBiomeType(biome, SAVANNA);
+            BiomeDictionary.addTypes(biome, SAVANNA);
         }
 
         if (biome.topBlock == Blocks.SAND)
         {
-            BiomeDictionary.registerBiomeType(biome, SANDY);
+            BiomeDictionary.addTypes(biome, SANDY);
         }
         else if (biome.topBlock == Blocks.MYCELIUM)
         {
-            BiomeDictionary.registerBiomeType(biome, MUSHROOM);
+            BiomeDictionary.addTypes(biome, MUSHROOM);
         }
         if (biome.fillerBlock == Blocks.HARDENED_CLAY)
         {
-            BiomeDictionary.registerBiomeType(biome, MESA);
+            BiomeDictionary.addTypes(biome, MESA);
         }
     }
 
     //Internal implementation
     private static BiomeInfo getBiomeInfo(Biome biome)
     {
-        return biomeInfoMap.get(Biome.REGISTRY.getNameForObject(biome));
+        BiomeInfo info = biomeInfoMap.get(biome.getRegistryName());
+        if (info == null)
+        {
+            info = new BiomeInfo();
+            biomeInfoMap.put(biome.getRegistryName(), info);
+        }
+        return info;
     }
 
-    private static void checkRegistration(Biome biome)
+    /**
+     * Ensure that the given biome has been tagged with at least one type
+     * @param biome the biome
+     */
+    static void ensureTagged(Biome biome)
     {
-        if(!isBiomeRegistered(biome))
+        if (!hasTypes(biome))
         {
+            FMLLog.warning("Biome %s has not been tagged with any types, types will be assigned on a best-effort guess.");
```

Comment:
To avoid "tag", this could be "Biome %s does not have any types added, types will be assigned on a best-effort guess."